### PR TITLE
[RFC] Support the latest versions of GHC

### PR DESCRIPTION
**Pull request by friedbrice**
This is a tentative PR. I'd like to merge it, but I'm not adamant. This PR eliminates almost all CPP, closing https://github.com/scotty-web/scotty/issues/252. (Some CPP is kept around for gating certain tests by platform support.)

So, I'd like to tackle some of the Issues, such as https://github.com/scotty-web/scotty/issues/262, https://github.com/scotty-web/scotty/issues/258, and https://github.com/scotty-web/scotty/issues/272, and I would need to smooth out and simplify project maintenance before doing so.

I'd like to hear what thoughts @scotty-web/admins, @fumieval, @RyanGlScott, and other interested parties have about the following changes.

* remove most CPP
* use Cabal 3.0
* add ghc 9.2, 9.4 to ci
* drop ghc pre 8.10 from ci
* move library files to src/
* fourmolu and hlint
* make copyright in license files agree with copyright in cabal files

### CODE REVIEW

File: Web/Scotty/Internal/Types.hs
Comment (friedbrice) [inline Web/Scotty/Internal/Types.hs:10]: Moved to src/Web/Scotty/Internal/Types.hs

File: bench/Main.hs
Comment (friedbrice) [inline bench/Main.hs:4]: Code formatter/linter. No functional changes.

File: examples/basic.hs
Comment (friedbrice) [inline examples/basic.hs:3]: Code formatter/linter. No functional changes.

File: examples/bodyecho.hs
Comment (friedbrice) [inline examples/bodyecho.hs:3]: Code formatter/linter. No functional changes.

File: examples/cookies.hs
Comment (friedbrice) [inline examples/cookies.hs:5]: Code formatter/linter. No functional changes.

File: examples/exceptions.hs
Comment (friedbrice) [inline examples/exceptions.hs:5]: Code formatter/linter. No functional changes.

File: examples/globalstate.hs
Comment (friedbrice) [inline examples/globalstate.hs:1]: Code formatter/linter. No functional changes.

File: examples/gzip.hs
Comment (friedbrice) [inline examples/gzip.hs:3]: Code formatter/linter. No functional changes.

File: examples/options.hs
Comment (friedbrice) [inline examples/options.hs:3]: Code formatter/linter. No functional changes.

File: examples/reader.hs
Comment (friedbrice) [inline examples/reader.hs:10]: Code formatter/linter. No functional changes.

File: examples/scotty-examples.cabal
Comment (friedbrice) [inline examples/scotty-examples.cabal:14]: Changed to Cabal 3.0

File: examples/scotty-examples.cabal
Comment (friedbrice) [inline examples/scotty-examples.cabal:16]: added GHC 9.4, 9.2. Dropped GHC <8.10

File: examples/scotty-examples.cabal
Comment (friedbrice) [inline examples/scotty-examples.cabal:29]: I put Scotty and GHC boot libs in a common `build-depends`. This adds some spurious `build-depends` to some of the examples, but I figure it's a moot point as there's not really going to be a Haskell distribution that doesn't include the GHC boot libs already.

File: examples/upload.hs
Comment (friedbrice) [inline examples/upload.hs:3]: Code formatter/linter. No functional changes.

File: examples/urlshortener.hs
Comment (friedbrice) [inline examples/urlshortener.hs:3]: Code formatter/linter. No functional changes.

File: scotty.cabal
Comment (friedbrice) [inline scotty.cabal:14]: Changed to Cabal 3.0

File: scotty.cabal
Comment (friedbrice) [inline scotty.cabal:47]: Add GHCs 9.4, 9.2. Drop GHCs <8.10

File: scotty.cabal
Comment (friedbrice) [inline scotty.cabal:67]: cosmetic refactor, to reduce future diff noise

File: scotty.cabal
Comment (friedbrice) [inline scotty.cabal:75]: I'm kinda skeptical that these bounds really make a lot of sense. How were they calculated?

In the PR, I only bound `base` and leave it up to Cabal. To compensate for lack of bounds in the cabal file, I've included freeze files from some successful builds with a variety of GHCs. If people are adamant about it, I can use the feeze files to compute version bounds.

File: src/Web/Scotty.hs
Comment (friedbrice) [inline src/Web/Scotty.hs:7]: Code formatter/linter. No functional change.

File: src/Web/Scotty/Action.hs
Comment (friedbrice) [inline src/Web/Scotty/Action.hs:75]: I don't necessarily agree with the code formatter, but I think it's useful for reducing future diff noise. I'm open to rolling back the formatting changes if people disagree.

File: src/Web/Scotty/Action.hs
Comment (friedbrice) [inline src/Web/Scotty/Action.hs:279]: I cleaned up this block by changing nested if/then to guards, and factored out `true` and `false` so that a sufficiently-advanced compiler could float them if it wanted.

File: src/Web/Scotty/Route.hs
Comment (friedbrice) [inline src/Web/Scotty/Route.hs:5]: * remove legacy CPP
* Code formatter/linter. No functional change.

File: src/Web/Scotty/Trans.hs
Comment (friedbrice) [inline src/Web/Scotty/Trans.hs:13]: Code formatter/linter. No functional change.

File: src/Web/Scotty/Util.hs
Comment (friedbrice) [inline src/Web/Scotty/Util.hs:12]: Code formatter/linter. No functional change.

File: src/Web/Scotty/Action.hs
Comment (friedbrice) [inline src/Web/Scotty/Action.hs:4]: * Remove legacy CPP.
* Code formatter/linter.

File: test/Web/ScottySpec.hs
Comment (friedbrice) [inline test/Web/ScottySpec.hs:187]: I moved some things that were top-level and nested them into the test block. Reads better with the CPP that way.

File: test/Web/ScottySpec.hs
Comment (friedbrice) [inline test/Web/ScottySpec.hs:4]: minor functional change by moving some helpers from top-level into a `let` bind. This looks like a huge diff, but set github to ignore whitespace, as the changes are mostly indentation.

File: src/Web/Scotty/Internal/Types.hs
Comment (friedbrice) [inline src/Web/Scotty/Internal/Types.hs:169]: Deriving `Semigroup`, `Monoid`, and `Monad`.

File: src/Web/Scotty/Internal/Types.hs
Comment (friedbrice) [inline src/Web/Scotty/Internal/Types.hs:174]: Moved implementations from `MonadPlus` instance to `Alternative` instance.

File: src/Web/Scotty/Internal/Types.hs
Comment (friedbrice) [inline src/Web/Scotty/Internal/Types.hs:9]: * moved from ./Web/Scotty/Internal/Types.hs.
* remove CPP
* derive (functionally compatible) instances instead of writing them

**Comment by fumieval** [comment]
It might be too late, but it would be nice to separate code style changes at least

**Comment by friedbrice** [comment]
@fumieval good call. i can easily split them out :-)

File: src/Web/Scotty/Internal/Types.hs
Comment (ocramz) [reply src/Web/Scotty/Internal/Types.hs:169]: doesn't this rule out old GHCs? or are we dropping them (which is also fine)

Comment (fumieval) [reply src/Web/Scotty/Internal/Types.hs:169]: > drop ghc pre 8.10 from ci

I suppose this is intentional

**Comment by jberryman** [comment]
would love to see 9.4 support merged (meantime 9.6 is out as well). Is there anything holding this up?

**Comment by ocramz** [comment]
> good call. i can easily split them out :-)

Hi @friedbrice , would you be able to move this forward so we do one last review and merge? Thank you!

File: freezes/ghc8107-2022-11-19-cabal.project.freeze
Comment (ocramz) [inline freezes/ghc8107-2022-11-19-cabal.project.freeze:1]: @friedbrice is this file necessary?

File: scotty.cabal
Comment (ocramz) [inline scotty.cabal:11]: there are other maintainers too at present

**Review by ocramz — APPROVED** [review]
Took me a while to review it all but eventually LGTM.

For the future let's not have formatting PRs mixed with code changes.

**Review by ocramz — CHANGES_REQUESTED** [review]
Apologies, I'm not able to fix the conflicts so I cannot merge these changes.

**Comment by ocramz** [comment]
Let's have separate PRs for linting and functional changes in the future.

**Pull request closed without merging**
